File: java/src/test/java/io/github/xesam/linkseal/TemplateEngineTest.java
package io.github.xesam.linkseal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import io.github.xesam.linkseal.resolver.TemplateEngine;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

final class TemplateEngineTest {

  private final TemplateEngine engine = new TemplateEngine();

  @Test
  void expandsSimpleVariableSubstitution() {
    assertEquals(
        "https://example.com/users/user_123",
        engine.expand("https://example.com/users/{userId}", Map.of("userId", "user_123"))
    );
  }

  @Test
  void expandsMultipleSimpleSubstitutions() {
    assertEquals(
        "/myorg/myrepo/issues/42",
        engine.expand("/{org}/{repo}/issues/{id}", Map.of("org", "myorg", "repo", "myrepo", "id", "42"))
    );
  }

  @Test
  void expandsQueryStartOperator() {
    assertEquals(
        "https://api.example.com/users/user_123?token=sess_abc&locale=en-US",
        engine.expand(
            "https://api.example.com/users/{userId}{?token,locale}",
            Map.of("userId", "user_123", "token", "sess_abc", "locale", "en-US")
        )
    );
  }

  @Test
  void expandsQueryContinuationOperator() {
    assertEquals(
        "https://api.example.com/users/user_123?source=app&token=sess_abc&locale=en-US",
        engine.expand(
            "https://api.example.com/users/{userId}?source=app{&token,locale}",
            Map.of("userId", "user_123", "token", "sess_abc", "locale", "en-US")
        )
    );
  }

  @Test
  void skipsMissingVariablesInQueryExpansion() {
    assertEquals(
        "/search?q=test",
        engine.expand("/search{?q,page}", Map.of("q", "test"))
    );
  }

  @Test
  void returnsEmptyWhenAllQueryVariablesMissing() {
    assertEquals("/search", engine.expand("/search{?q,page}", Map.of()));
  }

  @Test
  void simpleVariableMissingUsesEmptyString() {
    assertEquals("/users/", engine.expand("/users/{id}", Map.of()));
  }

  @Test
  void encodesSpecialCharacters() {
    assertEquals(
        "/search?q=hello%20world",
        engine.expand("/search{?q}", Map.of("q", "hello world"))
    );
  }

  // Mirrored byte-for-byte by the JavaScript SDK; these expected strings are the parity pin.
  // RFC 3986 unreserved = A-Za-z0-9-._~ passes through; everything else is %XX UTF-8.
  @Test
  void encodesFormStyleValuesPerRfc3986Unreserved() {
    assertEquals("/x?q=~", engine.expand("/x{?q}", Map.of("q", "~")));
    assertEquals("/x?q=%28", engine.expand("/x{?q}", Map.of("q", "(")));
    assertEquals("/x?q=%29", engine.expand("/x{?q}", Map.of("q", ")")));
    assertEquals("/x?q=%27", engine.expand("/x{?q}", Map.of("q", "'")));
    assertEquals("/x?q=%2A", engine.expand("/x{?q}", Map.of("q", "*")));
    assertEquals("/x?q=%2F", engine.expand("/x{?q}", Map.of("q", "/")));
    assertEquals("/x?q=%3F", engine.expand("/x{?q}", Map.of("q", "?")));
    assertEquals("/x?q=%23", engine.expand("/x{?q}", Map.of("q", "#")));
    assertEquals("/x?q=%21", engine.expand("/x{?q}", Map.of("q", "!")));
    assertEquals("/x?q=%24", engine.expand("/x{?q}", Map.of("q", "$")));
    assertEquals("/x?q=%26", engine.expand("/x{?q}", Map.of("q", "&")));
    assertEquals("/x?q=%2C", engine.expand("/x{?q}", Map.of("q", ",")));
    assertEquals("/x?q=%3B", engine.expand("/x{?q}", Map.of("q", ";")));
    assertEquals("/x?q=%3D", engine.expand("/x{?q}", Map.of("q", "=")));
    assertEquals("/x?q=%20", engine.expand("/x{?q}", Map.of("q", " ")));
    assertEquals("/x?q=%E4%B8%AD", engine.expand("/x{?q}", Map.of("q", "中")));
    assertEquals("/x?q=-._~", engine.expand("/x{?q}", Map.of("q", "-._~")));
    assertEquals("/x?q=a%20b%2Fc%28d%29e%2Af%27g", engine.expand("/x{?q}", Map.of("q", "a b/c(d)e*f'g")));
  }

  @Test
  void simpleExpansionUsesSameRfc3986Encoder() {
    assertEquals("/p/a%2Ab", engine.expand("/p/{v}", Map.of("v", "a*b")));
    assertEquals("/p/a~b", engine.expand("/p/{v}", Map.of("v", "a~b")));
    assertEquals("/p/%E4%B8%AD", engine.expand("/p/{v}", Map.of("v", "中")));
  }

  @Test
  void encodesDefinedEmptyValueAsNameEquals() {
    assertEquals("/x?q=", new TemplateEngine().expand("/x{?q}", Map.of("q", "")));
  }

  @Test
  void variableNamesExtractsSimplePlaceholders() {
    assertEquals(List.of("org", "repo", "id"), engine.variableNames("/{org}/{repo}/issues/{id}"));
  }

  @Test
  void variableNamesExtractsAcrossQueryOperators() {
    assertEquals(
        List.of("userId", "token", "locale"),
        engine.variableNames("https://api.example.com/users/{userId}?source=app{&token,locale}")
    );
  }

  @Test
  void variableNamesExtractsFromQueryStartOperator() {
    assertEquals(List.of("q", "page"), engine.variableNames("/search{?q,page}"));
  }

  @Test
  void variableNamesReturnsEmptyForTemplateWithoutPlaceholders() {
    assertEquals(List.of(), engine.variableNames("https://example.com/static"));
  }

  @Test
  void variableNamesDeduplicatesRepeatedName() {
    assertEquals(List.of("id"), engine.variableNames("/{id}/detail/{id}"));
  }

  @Test
  void variableNamesYieldsExactlyWhatExpandConsumes() {
    String template = "https://api.example.com/users/{userId}?source=app{&token,locale}";
    Map<String, String> values = new LinkedHashMap<>();
    for (String name : engine.variableNames(template)) {
      values.put(name, "x");
    }

    assertEquals("https://api.example.com/users/x?source=app&token=x&locale=x", engine.expand(template, values));
  }

  @Test
  void expandsCommaJoinedSimpleSubstitution() {
    assertEquals("foo,bar", engine.expand("{a,b}", Map.of("a", "foo", "b", "bar")));
  }

  // --- rejects unsupported RFC 6570 features (mirrored in the JS SDK) ---

  @Test
  void rejectsReservedExpansionOperator() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{+path}", Map.of("path", "/x")));
  }

  @Test
  void rejectsFragmentExpansionOperator() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{#frag}", Map.of("frag", "x")));
  }

  @Test
  void rejectsLabelExpansionOperator() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{.x}", Map.of("x", "1")));
  }

  @Test
  void rejectsPathExpansionOperator() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{/x}", Map.of("x", "1")));
  }

  @Test
  void rejectsPathStyleParameterOperator() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{;x}", Map.of("x", "1")));
  }

  @Test
  void rejectsExplodeModifier() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{?list*}", Map.of("list", "x")));
  }

  @Test
  void rejectsPrefixModifier() {
    assertThrows(IllegalArgumentException.class, () -> engine.expand("{?x:3}", Map.of("x", "abcdef")));
  }

  @Test
  void rejectsUnsupportedOperatorInVariableNames() {
    assertThrows(IllegalArgumentException.class, () -> engine.variableNames("{+path}"));
  }
}

File: javascript/packages/linkseal-resolver/src/__tests__/TemplateEngine.test.ts
import { describe, it, expect } from 'vitest';
import { TemplateEngine } from '../TemplateEngine.js';

describe('TemplateEngine', () => {
  it('expands simple variable substitutions', () => {
    const engine = new TemplateEngine();
    const result = engine.expand('https://example.com/users/{userId}', {
      userId: 'user_123',
    });
    expect(result).toBe('https://example.com/users/user_123');
  });

  it('expands multiple simple substitutions', () => {
    const engine = new TemplateEngine();
    const result = engine.expand('/{org}/{repo}/issues/{id}', {
      org: 'myorg',
      repo: 'myrepo',
      id: '42',
    });
    expect(result).toBe('/myorg/myrepo/issues/42');
  });

  it('expands {?} query start operator', () => {
    const engine = new TemplateEngine();
    const result = engine.expand(
      'https://api.example.com/users/{userId}{?token,locale}',
      { userId: 'user_123', token: 'sess_abc', locale: 'en-US' }
    );
    expect(result).toBe('https://api.example.com/users/user_123?token=sess_abc&locale=en-US');
  });

  it('expands {&} query continuation operator', () => {
    const engine = new TemplateEngine();
    const result = engine.expand(
      'https://api.example.com/users/{userId}?source=app{&token,locale}',
      { userId: 'user_123', token: 'sess_abc', locale: 'en-US' }
    );
    expect(result).toBe('https://api.example.com/users/user_123?source=app&token=sess_abc&locale=en-US');
  });

  it('skips missing variables in query expansion', () => {
    const engine = new TemplateEngine();
    const result = engine.expand(
      '/search{?q,page}',
      { q: 'test' }
    );
    expect(result).toBe('/search?q=test');
  });

  it('returns empty string when all query variables are missing', () => {
    const engine = new TemplateEngine();
    const result = engine.expand(
      '/search{?q,page}',
      {}
    );
    expect(result).toBe('/search');
  });

  it('replaces template variable with empty string when missing', () => {
    const engine = new TemplateEngine();
    const result = engine.expand('/users/{id}', {});
    expect(result).toBe('/users/');
  });

  it('handles URL encoding of special characters', () => {
    const engine = new TemplateEngine();
    const result = engine.expand('/search{?q}', { q: 'hello world' });
    expect(result).toBe('/search?q=hello%20world');
  });

  it('handles the spec E2E template', () => {
    const engine = new TemplateEngine();
    const result = engine.expand(
      'https://api.example.com/users/{userId}?source=app{&token,locale}',
      { userId: 'user_123', token: 'sess_abc', locale: 'en-US' }
    );
    expect(result).toBe('https://api.example.com/users/user_123?source=app&token=sess_abc&locale=en-US');
  });

  it('expands comma-joined simple substitution', () => {
    const engine = new TemplateEngine();
    const result = engine.expand('{a,b}', { a: 'foo', b: 'bar' });
    expect(result).toBe('foo,bar');
  });

  describe('rejects unsupported RFC 6570 features', () => {
    const engine = new TemplateEngine();

    it('rejects reserved expansion {+}', () => {
      expect(() => engine.expand('{+path}', { path: '/x' })).toThrow('Unsupported RFC 6570 operator');
    });

    it('rejects fragment expansion {#}', () => {
      expect(() => engine.expand('{#frag}', { frag: 'x' })).toThrow('Unsupported RFC 6570 operator');
    });

    it('rejects label expansion {.}', () => {
      expect(() => engine.expand('{.x}', { x: '1' })).toThrow('Unsupported RFC 6570 operator');
    });

    it('rejects path expansion {/}', () => {
      expect(() => engine.expand('{/x}', { x: '1' })).toThrow('Unsupported RFC 6570 operator');
    });

    it('rejects path-style parameter expansion {;}', () => {
      expect(() => engine.expand('{;x}', { x: '1' })).toThrow('Unsupported RFC 6570 operator');
    });

    it('rejects the explode modifier *', () => {
      expect(() => engine.expand('{?list*}', { list: 'x' })).toThrow('Unsupported modifier');
    });

    it('rejects the prefix modifier :N', () => {
      expect(() => engine.expand('{?x:3}', { x: 'abcdef' })).toThrow('Unsupported modifier');
    });

    it('rejects in variableNames too', () => {
      expect(() => engine.variableNames('{+path}')).toThrow('Unsupported RFC 6570 operator');
    });
  });

  describe('value encoding (strict RFC 3986 unreserved)', () => {
    // Mirrored byte-for-byte by the Java SDK; these expected strings are the parity pin.
    // RFC 3986 unreserved = A-Za-z0-9-._~ passes through; everything else is %XX UTF-8.
    const cases: Array<[string, string]> = [
      ['~', '?q=~'],
      ['(', '?q=%28'],
      [')', '?q=%29'],
      ["'", '?q=%27'],
      ['*', '?q=%2A'],
      ['/', '?q=%2F'],
      ['?', '?q=%3F'],
      ['#', '?q=%23'],
      ['!', '?q=%21'],
      ['$', '?q=%24'],
      ['&', '?q=%26'],
      [',', '?q=%2C'],
      [';', '?q=%3B'],
      ['=', '?q=%3D'],
      [' ', '?q=%20'],
      ['中', '?q=%E4%B8%AD'],
      ['-._~', '?q=-._~'],
      ["a b/c(d)e*f'g", '?q=a%20b%2Fc%28d%29e%2Af%27g'],
    ];

    it.each(cases)('form-style {?q} encodes %j', (value, expected) => {
      expect(new TemplateEngine().expand('/x{?q}', { q: value })).toBe('/x' + expected);
    });

    it('simple expansion {v} uses the same encoder', () => {
      const engine = new TemplateEngine();
      expect(engine.expand('/p/{v}', { v: 'a*b' })).toBe('/p/a%2Ab');
      expect(engine.expand('/p/{v}', { v: 'a~b' })).toBe('/p/a~b');
      expect(engine.expand('/p/{v}', { v: '中' })).toBe('/p/%E4%B8%AD');
    });

    it('encodes a defined empty value as name= (not skipped)', () => {
      expect(new TemplateEngine().expand('/x{?q}', { q: '' })).toBe('/x?q=');
    });
  });

  describe('variableNames', () => {
    it('extracts simple placeholders', () => {
      expect(new TemplateEngine().variableNames('/{org}/{repo}/issues/{id}')).toEqual([
        'org',
        'repo',
        'id',
      ]);
    });

    it('extracts names across query operators', () => {
      expect(
        new TemplateEngine().variableNames(
          'https://api.example.com/users/{userId}?source=app{&token,locale}'
        )
      ).toEqual(['userId', 'token', 'locale']);
    });

    it('extracts names from the query start operator', () => {
      expect(new TemplateEngine().variableNames('/search{?q,page}')).toEqual(['q', 'page']);
    });

    it('returns an empty list for a template with no placeholders', () => {
      expect(new TemplateEngine().variableNames('https://example.com/static')).toEqual([]);
    });

    it('deduplicates a name used more than once', () => {
      expect(new TemplateEngine().variableNames('/{id}/detail/{id}')).toEqual(['id']);
    });

    it('yields exactly the names expand consumes', () => {
      const engine = new TemplateEngine();
      const template = 'https://api.example.com/users/{userId}?source=app{&token,locale}';
      const values = Object.fromEntries(engine.variableNames(template).map((n) => [n, 'x']));

      expect(engine.expand(template, values)).toBe(
        'https://api.example.com/users/x?source=app&token=x&locale=x'
      );
    });
  });
});
